### ISSUE TITLE
timeLimitMultiplier compounds under module refinement

### ISSUE THREAD

**Issue by parno** [dafny-lang/dafny#1224]
Consider this pair of modules:
```
abstract module SlowPoke {
  lemma {:timeLimitMultiplier 4} SlowLemma()
}

module SlowChild refines SlowPoke { }
```
If we run Dafny with `/timeLimit 50` or `/rlimit 50`, we would expect the resulting Boogie attribute to be a timeLimit/rlimit of 200, but the procedures from `SlowChild` end up with 1000, which presumably is not what the developer intended.  

The code responsible https://github.com/dafny-lang/dafny/blob/b7f16e4cf1fb54636d077bfe21668881c5e093c4/Source/Dafny/Rewriter.cs#L1517 appears to update the original `timeLimitMultiplier` attribute in place (changing its name and value), rather than, say, adding a new attribute.  Looking at this in the debugger, when we first process `SlowPoke`, we find that the `timeLimitMultiplier` attribute `value` is 4 and the `current_limit` is 50.  However, when we process `SlowChild`, we find that the `timeLimitMultiplier` attribute `value` is 200 and the `current_limit` is 50. It's a bit puzzling why the `value` has changed but the `Name` has not.  

Perhaps this is a bug in the `RefinementTransformer`?  I'm tempted to say that when we're processing `timeLimitMulitiplier`, we should instead add a new attribute to the member, rather than trying to overwrite the existing one.  However, there's a comment in the code https://github.com/dafny-lang/dafny/blob/b7f16e4cf1fb54636d077bfe21668881c5e093c4/Source/Dafny/DafnyAst.cs#L3443 indicating that this field is supposed to be read-only.

@RustanLeino Any thoughts on the best way to address this?

**Comment by RustanLeino** [comment]
I agree this is a bug.

From running Dafny and looking at its `/rprint` output, it appears that `:timeLimitMultiplier` is applied once in `SlowPoke` and effectively twice in `SlowChild`. Furthermore, `:timeLimitMultiplier` is removed at some point (I didn't look into when, but it must be removed after the processing of `refines`, since it keeps multiplying).

Attributes in a Dafny program cannot start with an `_`. However, internally, Dafny uses attributes that start with an `_`. This is used in some places where the user writes `X` and Dafny processes attributes and adds an `_X` attribute (despite what that comment on L3443 seems to say). I suggest that to fix this issue, the machinery that looks at these limits and multipliers _not_ change any given `:timeLimit`/`:rlimit`, but instead adds a `:_timeLimit`/`:_rlimit` attribute when the result of the multiplication.

The refinement transformer (or any other cloner) should then not copy attributes that start with `_`, and thus `SlowChild` will end up with the same product as in `SlowPoke`.

Finally, when the Boogie program is generated, if there's a `:_timeLimit` (or, analogously for rlimit), then it gets used as the Boogie `:timeLimit` and any Dafny `:timeLimit` is ignored. If there is no `:_timeLimit`, then any `:timeLimit` gets copied into the Boogie.

In summary, I'm suggesting that instead of overwriting `:timeLimit`, the code ought to write a new `:_timeLimit`, and the `:_timeLimit` should take priority of `:timeLimit` in the translation to Boogie. Does that make sense?

**Comment by parno** [comment]
Yes, this makes sense.  I'll give it a try.  Thanks!
